crates/lib/src/execution_data.rs

```diff
@@ -1,58 +1,130 @@
+use enum_as_inner::EnumAsInner;
+use num::complex::{Complex32, Complex64};
 use std::collections::HashMap;
-use std::convert::TryFrom;
-use std::num::{ParseIntError, TryFromIntError};
+use std::num::ParseIntError;
 use std::str::FromStr;
 use std::time::Duration;
 
-use qcs_api_client_grpc::models::controller::ReadoutValues;
+use ndarray::prelude::*;
+
+use qcs_api_client_grpc::models::controller::{readout_values, ReadoutValues};
 use quil_rs::instruction::MemoryReference;
 
 use crate::RegisterData;
 
-/// The result of executing an [`Executable`](crate::Executable) via
-/// [`Executable::execute_on_qvm`](crate::Executable::execute_on_qvm).
+/// The result of executing an [`Executable`](crate::Executable)
 #[derive(Debug, Clone, PartialEq)]
-pub struct Qvm {
+pub struct ExecutionData {
     /// The readout data that was read from the [`Executable`](crate::Executable).
     /// Key is the name of the register, value is the data of the register after execution.
-    pub registers: HashMap<Box<str>, RegisterData>,
+    pub readout_data: ReadoutMap,
     /// The time it took to execute the program on the QPU, not including any network or queueing
     /// time. If paying for on-demand execution, this is the amount you will be billed for.
     ///
     /// This will always be `None` for QVM execution.
     pub duration: Option<Duration>,
 }
 
+/// An enum representing every possible readout type
+#[derive(Clone, Copy, Debug, EnumAsInner, PartialEq)]
+pub enum ReadoutTypes {
```

Comment:
Should we simplify this enum to something like `Integer(i32)`, `Real(f64)`, and `Complex(Complex64)`?

The benefits are a simpler API, especially if you are using this in Rust. The tradeoff is memory efficiency, but users that would be concerned with that kind of performance are likely using data from a real `QPU`, and that's already returned in the biggest possible types anyways.